tests/test_task_engine.py

```diff
@@ -38,6 +41,79 @@
 from prefect.utilities.engine import propose_state
 
 
+@pytest.fixture(autouse=True, params=[False, True])
+def enable_client_side_task_run_orchestration(
+    request, asserting_events_worker: EventsWorker
+):
+    enabled = request.param
+    with temporary_settings(
+        {PREFECT_EXPERIMENTAL_ENABLE_CLIENT_SIDE_TASK_ORCHESTRATION: enabled}
+    ):
+        yield enabled
+
+
+async def get_task_run(task_run_id: Optional[UUID]) -> TaskRun:
+    if PREFECT_EXPERIMENTAL_ENABLE_CLIENT_SIDE_TASK_ORCHESTRATION:
+        task_run = get_task_run_sync(task_run_id)
+    else:
+        client = get_client()
+        if task_run_id:
+            task_run = await client.read_task_run(task_run_id)
+        else:
+            task_runs = await client.read_task_runs()
+            task_run = task_runs[-1]
+
+    return task_run
+
+
+def get_task_run_sync(task_run_id: Optional[UUID]) -> TaskRun:
+    if PREFECT_EXPERIMENTAL_ENABLE_CLIENT_SIDE_TASK_ORCHESTRATION:
+        events = AssertingEventsClient.last.events
+        events = sorted(events, key=lambda e: e.occurred)
+        if task_run_id:
+            events = [
+                e
+                for e in events
+                if e.resource.prefect_object_id("prefect.task-run") == task_run_id
+            ]
+        last_event = events[-1]
+        state = State(**last_event.payload["validated_state"])
+        task_run = TaskRun(
+            id=last_event.resource.prefect_object_id("prefect.task-run"),
+            state=state,
+            state_id=state.id,
+            state_type=state.type,
+            state_name=state.name,
+            **last_event.payload["task_run"],
+        )
+    else:
+        client = get_client(sync_client=True)
+        if task_run_id:
+            task_run = client.read_task_run(task_run_id)
+        else:
+            task_runs = client.read_task_runs()
+            task_run = task_runs[-1]
+
+    return task_run
+
+
+async def get_task_run_states(task_run_id: UUID) -> List[State]:
+    if PREFECT_EXPERIMENTAL_ENABLE_CLIENT_SIDE_TASK_ORCHESTRATION:
```

Comment:
is there a flush method I can use for `asserting_events_worker` but not call `drain()` and stop the service? I got a flake that was missing one the events

Reply:
Is this still relevant with our refactoring to use the `AssertingEventsClient` directly?

Reply:
I only saw it actually after moving to `AssertingEventsClient`. I _think_ it was an instance where an item was put on the queue but we were checking `AssertingEventsClient` too soon

Reply:
Since the queueservice is fundamentally async from the engine, unless there's _some_ kind of checkpoint where we can guarantee it's drained, we'll end up with flakes here.  Maybe we should just give ourselves a way to drain it without stopping it, then use that in the test utilities like `get_task_run` and such

Reply:
thoughts on this? https://github.com/PrefectHQ/prefect/pull/14661/commits/c2204194492c43b84c00302a46d5118b84a730d4

little scared to even touch `_main_loop` but this seems harmless enough

Reply:
YAS